### PR TITLE
[ESQL] Migrate SimplifiyComparisonArithmetics optimization rule

### LINKED ISSUE #108388: [ESQL] Class cast exception mixing division and comparison
### Elasticsearch Version

main

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

n/a

### Problem Description

`FROM test | EVAL x = double / 2 > 4` results in a `ClassCastException`:

```
        "reason": "class org.elasticsearch.compute.data.DoubleVectorBlock cannot be cast to class org.elasticsearch.compute.data.IntBlock (org.elasticsearch.compute.data.DoubleVectorBlock and org.elasticsearch.compute.data.IntBlock are in unnamed module of loader java.net.FactoryURLClassLoader @45e11627)"
```

This seems to only happen when both division and comparison are involved.  Neither `FROM test | EVAL x = double / 2` and `FROM test | EVAL x = double > 4` trigger the error.

### Steps to Reproduce

```
PUT http://localhost:9200/test
content-type: application/json

{
  "mappings": {
    "properties": {
        "float": {"type": "float"},
        "int": {"type": "integer"},
        "double": {"type": "double"}
    }
  }
}


POST http://localhost:9200/_query?format=txt
content-type: application/json

{
  "version": "2024.04.01",
  "query": "FROM test | EVAL x = double / 2 > 4" 
}
```

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by not-napoleon**
This PR migrates the QL optimization rule `SimplifyComparisonArithmetics`, which folds constant arithmetic in binary comparisons when possible.  ~I didn't find any tests for this optimization rule in the existing code base, so I wrote a few that cover the cases I could think of.~  I also slightly refactored the optimization to not rely on string comparing the symbols to determine the operation.  That was necessary in the generic code, since the operations were in a different module, but by pulling this into the ESQL module, we have direct access to the relevant enum.  [This commit](https://github.com/elastic/elasticsearch/commit/50f4bb6660e3e009e84ad9d90817778bd1e4c285) isolates that refactoring, which might be easier to review.

I migrated the tests from `OptimizerRunTests` into `LogicalPlanOptimizerTests`.  As you can see, there were some significant difficulties in doing that.  I've opened issues and applied the `@AwaitsFix` label for now.  My recommendation is that we merge this first and fix those issues in a follow up, as this review is already quite lengthy and I'm hesitant to make more tricky changes in it.

In the course of migrating the tests for this, I ran into a lot of issues:
    - Resolves https://github.com/elastic/elasticsearch/issues/108388
    - https://github.com/elastic/elasticsearch/issues/108524
    - https://github.com/elastic/elasticsearch/issues/108525
    - https://github.com/elastic/elasticsearch/issues/108519

I was able to resolve the first one, by replacing the usage of `Literal.of` with directly calling `new Literal()`, as `of()` was setting the data type incorrectly.

### CODE REVIEW

**Review by astefan — COMMENTED** [review]
Thank you for adding the tests 👍 .

For reference, there is one test that's EQL specific from the [original PR](https://github.com/elastic/elasticsearch/pull/66022) here `org.elasticsearch.xpack.eql.optimizer.OptimizerTests.testReduceBinaryComparisons`.

But, there are many more tests here https://github.com/elastic/elasticsearch/pull/66022/files#diff-e76f0734cca452532a386eba427f1c473a72f507740086ac04db5a370f3fbc6f

And many many more here https://github.com/elastic/elasticsearch/pull/66022/files#diff-66ac572ee225421b48743da1a178df51ef4568b9ee965acdf39125ca845efca7. The `sql-spec` files test ES SQL against in-memory H2 database, that's why there are no results for those queries (the results from ES and H2 are compared with eachother).

**Comment by not-napoleon** [comment]
Thanks @astefan , I'll work on getting those tests migrated.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/OptimizerRules.java [hunk 1]
```diff
@@ -612,4 +630,216 @@ private static Expression propagate(Or or) {
...
+            Literal bcLiteral = (Literal) bc.right();
+            Expression literalNeg = tryFolding(new Neg(bcLiteral.source(), bcLiteral));
+            return literalNeg == null ? bc : bc.reverse().replaceChildren(asList(((Neg) bc.left()).field(), literalNeg));
+        }
+
+        private static Expression tryFolding(Expression expression) {
+            if (expression.foldable()) {
+                try {
+                    expression = new Literal(expression.source(), expression.fold(), expression.dataType());
+                } catch (ArithmeticException | DateTimeException e) {
```
Comment (not-napoleon) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/OptimizerRules.java:691]: I believe this is the source of the issue in https://github.com/elastic/elasticsearch/issues/108519, as ES|QL will never throw from folding.  It's not as simple as just checking if we got a non-null value in the literal here though, as generating the null literal will have put the warning on the stack already.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/OptimizerRules.java [hunk 2]
```diff
@@ -612,4 +630,216 @@ private static Expression propagate(Or or) {
...
+                    return true;
+                }
+
+                return isOpUnsafe();
+            }
+
+            final Expression apply() {
+                // force float point folding for FlP field
+                Literal bcl = operation.dataType().isRational()
+                    ? new Literal(bcLiteral.source(), ((Number) bcLiteral.value()).doubleValue(), DataTypes.DOUBLE)
```
Comment (not-napoleon) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/OptimizerRules.java:751]: This is changed from the original rule.  The original version read `? Literal.of(bcLiteral, ((Number) bcLiteral.value()).doubleValue())`, however `Literal.of()` takes the data type from the first parameter, not the second.  This was causing us to create (e.g.) `Literals` with a value of `4.0` and a type of `INTEGER`, which in turn caused a class cast exception.  See https://github.com/elastic/elasticsearch/issues/108388

**Comment by not-napoleon** [comment]
@astefan I finished migrating the tests from `OptimizerRunTests` (with `AwaitsFix` notations, as we discussed).  I would like to merge this now, without including the SQL spec tests, and add them in a follow up PR.  Migrating the spec tests will be a lot of work, as the tests as written assume the existence of a reference database implementation to validate against.  Migrating them will require first re-writing them into ESQL, then working out the expected results for them.  This PR already contains one real bugfix, and there's several follow up bug fixes that we can't really address until this merges, and I'd rather not delay landing those.  What do you think?

**Comment by not-napoleon** [comment]
After discussion on the general approach to this migration, I am closing this PR in favor of https://github.com/elastic/elasticsearch/pull/108744, which migrates the SQL tests, and the first step of https://github.com/elastic/elasticsearch/issues/106679 which will pull over the relevant optimization rule.

**Pull request closed without merging**